### PR TITLE
Adding Support for Uint8arrays in createTypedArrays

### PR DESCRIPTION
**Pull request by Charlyjcmu**
We have added support for Uint8arrays when using IndexDatatype.createTypedArrays in IndexDatatype.js for issue #10349 . We have added a numberOfVertices check inside of the function and created a new constant called TWO_FIFTY_SIX_BYTES inside Math.js

We have corrected a few test cases inside Core/Geometry/Pipeline, Core/IndexDatatype, Scene/PrimitivePipeline, Scene/Model/GeoJsonLoader.

We are a little confused on how to \ correct the render test cases, because there is an internal GL_INVALID_OPERATION: Insufficient buffer size error. Most of the errors are just expecting to render some pixel but not rendering it.

We would like some help on how to get started correcting the following test cases:
- [ ] Scene/Geometry3DTileContent
- [ ] Scene/Vector3DTileContent
- [ ] Scene/Vector3DTileGeometry
- [ ] Scene/Vector3DTilePolygons
- [ ] Scene/Vector3DTilePolylines

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @Charlyjcmu!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/main/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/main/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.

**Comment by ggetz** [comment]
@j9liu As you opened the original issue, would you mind providing some guidance on the tests here? Particularly around the error:

> `DeveloperError: IndexDatatype.UNSIGNED_INT requires OES_element_index_uint, which is not supported on this system.  Check context.elementIndexUint.`

If this requires a particular extension, should there be a fall back in place for systems that its not supported on?

File: packages/engine/Source/Core/Math.js [hunk 1]
```diff
@@ -183,6 +183,13 @@ CesiumMath.SOLAR_RADIUS = 6.955e8;
  */
 CesiumMath.LUNAR_RADIUS = 1737400.0;
 
+/**
+ * 256
+ * @type {Number}
+ * @constant
+ */
+CesiumMath.TWO_FIFTY_SIX_BYTES = 256;
```
Comment (j9liu) [inline packages/engine/Source/Core/Math.js:191]: Keep in mind that `IndexDatatype.UNSIGNED_BYTE` corresponds to just one byte, and one byte = 256 bits.
```suggestion
CesiumMath.TWO_FIFTY_SIX_BITS = 256;
```

**Comment by j9liu** [comment]
@ggetz I'll take a look at this PR but I am confused why the `UNSIGNED_INT` would be causing errors now, since this PR is just adding support for typed arrays with uint8 or `UNSIGNED_BYTE`. Was there no fallback in the first place for using `UNSIGNED_INT` ?

It also seems like this functionality is provided by WebGL2, and [only in WebGL1 is the extension needed](https://developer.mozilla.org/en-US/docs/Web/API/OES_element_index_uint). So what would the correct procedure be for checking `UNSIGNED_INT` support when #10894 is merged?

File: packages/engine/Source/Core/IndexDatatype.js [hunk 2]
```diff
@@ -132,8 +132,11 @@ IndexDatatype.createTypedArray = function (
   if (numberOfVertices >= CesiumMath.SIXTY_FOUR_KILOBYTES) {
     return new Uint32Array(indicesLengthOrArray);
   }
+  if (numberOfVertices >= CesiumMath.TWO_FIFTY_SIX_BYTES) {
```
Comment (j9liu) [inline packages/engine/Source/Core/IndexDatatype.js:135]: I didn't note this when I wrote the issue, but this change should probably apply to `createTypedArrayFromArrayBuffer` as well (the function below this one)

File: packages/engine/Source/Core/IndexDatatype.js [hunk 3]
```diff
@@ -109,12 +109,12 @@ IndexDatatype.validate = function (indexDatatype) {
 };
 
 /**
- * Creates a typed array that will store indices, using either <code><Uint16Array</code>
- * or <code>Uint32Array</code> depending on the number of vertices.
+ * Creates a typed array that will store indices, using either <code><Uint8Array</code>
+ * or <code><Uint16Array</code> or <code>Uint32Array</code> depending on the number of vertices.
```
Comment (j9liu) [inline packages/engine/Source/Core/IndexDatatype.js:113]: Minor readability tweak:
```suggestion
 * Creates a typed array that will store indices, using either <code><Uint8Array</code>,
 * <code><Uint16Array</code>, or <code>Uint32Array</code> depending on the number of vertices.
```

File: packages/engine/Source/Core/IndexDatatype.js [hunk 4]
```diff
@@ -109,12 +109,12 @@ IndexDatatype.validate = function (indexDatatype) {
...
  *
  * @param {Number} numberOfVertices Number of vertices that the indices will reference.
  * @param {Number|Array} indicesLengthOrArray Passed through to the typed array constructor.
- * @returns {Uint16Array|Uint32Array} A <code>Uint16Array</code> or <code>Uint32Array</code> constructed with <code>indicesLengthOrArray</code>.
+ * @returns {Uint8Array|Uint16Array|Uint32Array} A <code>Uint8Array</code> or <code>Uint16Array</code> or <code>Uint32Array</code> constructed with <code>indicesLengthOrArray</code>.
```
Comment (j9liu) [inline packages/engine/Source/Core/IndexDatatype.js:117]:
```suggestion
 * @returns {Uint8Array|Uint16Array|Uint32Array} A <code>Uint8Array</code>, <code>Uint16Array</code>, or <code>Uint32Array</code> constructed with <code>indicesLengthOrArray</code>.
```

**Comment by ggetz** [comment]
Thanks @j9liu!

> Was there no fallback in the first place for using UNSIGNED_INT ?

`Context.elementIndexUint` has a check in place already, but bases on the test result, my guess is that value is not being properly accounted for.

> It also seems like this functionality is provided by WebGL2, and [only in WebGL1 is the extension needed](https://developer.mozilla.org/en-US/docs/Web/API/OES_element_index_uint). So what would the correct procedure be for checking UNSIGNED_INT support when https://github.com/CesiumGS/cesium/pull/10894 is merged? 

It will still be possible to opt into WebGL1, in which case we should still be checking for extension support. The above function should already account for that.

**Review by j9liu — COMMENTED** [review]
The problem with the unit tests comes from rendering problems with the geometry itself. The changes have caused holes in some of the geometry:

| main | this branch |
| ---- | -------------|
| image | image |

There might be code in the geometry classes that assumes the size of the indices being output from `createTypedArrays`. I'll have to do a sweep through the code for mentions of `IndexDatatype.createTypedArrays` and see where the problem areas are.

**Comment by j9liu** [comment]
I'll have to return to this later but from snooping around I found a couple things:

- There's this part of the code that has a `Uint16` vs `Uint32` bit check that we missed
https://github.com/CesiumGS/cesium/blob/27abb992eb5a434f1c9391f5e926643473ebc985/packages/engine/Source/Renderer/VertexArray.js#L678-L697
- This is what happens in the 3D Tiles Terrain Classification sandcastle (which uses `Vector3DTileContent`). This also happens in the ArcticDEM sandcastle.
image
- Something I noticed about `EllipseGeometry` while trying to debug it: it looks like both `Uint16` and `Uint8` arrays get made at some point in the progress. So if these differently-sized indices get packed together, they're going to lead to bad results. We'd need to look closer into that code though to figure out what incorrect assumptions are being made.

**Comment by j9liu** [comment]
Hi @Charlyjcmu,

Were you and your team going to continue work on this? Even though it has become more difficult than expected, I'm willing to investigate the areas that are producing these bugs and provide guidance on what needs changing.

Otherwise, this PR will be closed for inactivity and I will update the issue to describe the full scope of this problem.

**Comment by Charlyjcmu** [comment]
Yes, our team will keep working on it. We took a break for a while, but we will keep working on it for a while longer.

**Pull request closed without merging**